### PR TITLE
Inital mockup for rotations created

### PR DESCRIPTION
**Pull request by TheCelavi**

### CODE REVIEW

File: Destination/Destination.php [hunk 1]
```diff
@@ -3,15 +3,25 @@
...
 
 use Psr\Log\LoggerInterface;
+use \Countable;
+use \IteratorAggregate;
 
 /**
  * @author Kevin Bond <kevinbond@gmail.com>
  */
-interface Destination
+interface Destination extends Countable, IteratorAggregate
```
Comment (kbond) [inline Destination/Destination.php]: Having the destination interface `Countable`/`IteratorAggregate` seems strange to me. It makes it seem like there are multiple desinations when in fact, there are multiple Backups.

**Comment by kbond** [comment]
I do like the system but I think we should leave the existing `Destination` interface as-is and add another interface called `RotatingDestination` which extends `Destination`:
1. It doesn't make sense to have the `S3CmdDestination` rotatable.
2. I think we should have a `StreamDestination` and a `RotatingStreamDestination`. It is my opinion that most use cases will not require rotation.

File: Destination/Destination.php (hunk 1 above)
Comment (TheCelavi) [inline Destination/Destination.php]: It is matter of coding style, destination holds collection of backups, so it is aggregating object.

However, it is your library, we can replace that with method 'getBackups' and achieve same result.

I am for any approach, booth are correct and I do not mind to follow coding style of owner - if you do not like something, do yell :-) no problem.

N.

**Comment by TheCelavi** [comment]
Like! Even better!

File: Destination/Destination.php (hunk 1 above)
Comment (kbond) [inline Destination/Destination.php]: > destination holds collection of backups, so it is aggregating object

Backups may not be all it holds. We could have a `MirrorStreamDestination`, does `count($destination)` count the backups or the mirrors? Meh, probably just splitting hairs but if there is no great reason to have it, let's leave it out.

File: Destination/AbstractDestination.php [hunk 2]
```diff
@@ -0,0 +1,170 @@
+<?php
+
+namespace Zenstruck\BackupBundle\Destination;
+
+use Psr\Log\LoggerInterface;
+use Zenstruck\BackupBundle\Rotator\Rotator;
+
+abstract class AbstractDestination implements Destination
```
Comment (kbond) [inline Destination/AbstractDestination.php]: I try not using `Abstract` to prefix abstract classes if I can avoid it. Can we call this something else?

What about an abstract class called `RotatingDestination` instead of the interface? (not sure if I like not using an interface though...)

File: Destination/Destination.php (hunk 1 above)
Comment (TheCelavi) [inline Destination/Destination.php]: Hahahahahahhaha ! Awesome ! 

You are absolutely right!

But we can have master destination, slave/mirror destinations. We list from the master, slaves gets synced. So MirroredDestination can implement same Destination interface, with additional methods:
- getMaster()
- getMirrors()
- getAll()

Underlying implementation will take care that mirrors are up to date with master.

Love the idea about mirroring! Love it! I totally forgot about that, it is really useful for paranoid sys admins :-)

File: Destination/AbstractDestination.php (hunk 2 above)
Comment (TheCelavi) [inline Destination/AbstractDestination.php]: Again, it is a coding style. I prefer to have prefix/suffix when it comes to abstract classes, interfaces and traits.

Your proposal is fine, I will change the name.

File: Destination/Destination.php [hunk 3]
```diff
@@ -3,15 +3,25 @@
...
      */
     public function push($filename, LoggerInterface $logger);
+
+    /**
+     * Remove backup from backup destination.
+     *
+     * @param Backup $backup
+     * @param LoggerInterface $logger
+     */
+    public function remove(Backup $backup, LoggerInterface $logger);
```
Comment (kbond) [inline Destination/Destination.php:26]: I am not seeing the reason for `remove` to be a part of the public API.

**Comment by kbond** [comment]
The more I think about it, I don't even see a reason for a `RotatingDestination` interface.

**Comment by kbond** [comment]
I'm thinking all you need is:

`RotatingDestination` (abstract, implements `Destination`)
`RotatingStreamDestination` (extends `RotatingDestination`)

File: Destination/StreamDestination.php [hunk 4]
```diff
@@ -7,22 +7,49 @@
 /**
  * @author Kevin Bond <kevinbond@gmail.com>
  */
-class StreamDestination implements Destination
+class StreamDestination extends AbstractDestination
 {
     private $directory;
 
-    public function __construct($directory)
+    public function __construct($directory, array $preRotators = [], array $postRotators = [])
```
Comment (kbond) [inline Destination/StreamDestination.php]: Can you use php 5.3 array syntax to keep compatibility?

Comment (TheCelavi) [inline Destination/StreamDestination.php]: Of course.

File: Rotator/MaxStorageSizeRotator.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+     * @var integer
+     */
+    private $maxSize;
+
+    /**
+     * A constructor.
+     *
+     * @param integer $maxSize Maximum storage size.
+     */
+    public function __construct($maxSize)
```
Comment (kbond) [inline Rotator/MaxStorageSizeRotator.php:20]: Let's allow pretty sizes like "4G" (see https://github.com/zendframework/ZendDiagnostics/blob/master/src/ZendDiagnostics/Check/DiskFree.php#L215)

File: Destination/Destination.php (hunk 3 above)
Comment (TheCelavi) [inline Destination/Destination.php:26]: Ok, will see that after implementing modification and I will get back to you.

File: Rotator/MaxStorageSizeRotator.php (hunk 5 above)
Comment (TheCelavi) [inline Rotator/MaxStorageSizeRotator.php:20]: Agree.

Comment (TheCelavi) [inline Rotator/MaxStorageSizeRotator.php:20]: I will add Utils class with that static method, no need for dependency.

Comment (kbond) [inline Rotator/MaxStorageSizeRotator.php:20]: Maybe this lib can be an optional requirement: https://github.com/gabrielelana/byte-units

We could check if the value of $maxSize is an integer. If so, just use it. If not, ensure the package above is loaded and parse with it.

File: Destination/Backup.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        return $this->createdAt;
+    }
+
+    /**
+     * Builds backup from file.
+     *
+     * @param string $path Path to file.
+     * @return Backup Backup file.
+     */
+    public static function fromFile($path)
```
Comment (TheCelavi) [inline Destination/Backup.php:81]: Utility function added - builder pattern.

File: Destination/Destination.php (hunk 1 above)
Comment (TheCelavi) [inline Destination/Destination.php]: I do not know if you want me to remove Countable, IteratorAggregate, I have left those. Still think that it is not wrong approach - take per example Twig_Node implementation... It is common practice for aggregating object to implement iterator.

File: Destination/RotatableStreamDestination.php [hunk 7]
```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace Zenstruck\BackupBundle\Destination;
+
+use Psr\Log\LoggerInterface;
+use Zenstruck\BackupBundle\Rotator\Rotator;
+
+class RotatableStreamDestination extends StreamDestination
```
Comment (TheCelavi) [inline Destination/RotatableStreamDestination.php:8]: Ok, here it is, no new interface, no abstract class, only subclass of StreamDestination

File: Destination/RotatableStreamDestination.php [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    private $postRotator;
+
+    /**
+     * A constructor
+     *
+     * @param string $directory Path to backup directory.
+     * @param Rotator|null $preRotator Rotator to execute prior to push.
+     * @param Rotator|null $postRotator Rotator to execute after push.
+     */
+    public function __construct($directory, Rotator $preRotator = null, Rotator $postRotator = null)
```
Comment (TheCelavi) [inline Destination/RotatableStreamDestination.php:27]: I got the idea that you can provide instance of Rotator interface. If you have to use collection of rotators, there is a RotatorCollection that implements Rotator interface.

File: Destination/RotatableStreamDestination.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        parent::__construct($directory);
+        $this->preRotator = $preRotator;
+        $this->postRotator = $postRotator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function push($filename, LoggerInterface $logger)
```
Comment (TheCelavi) [inline Destination/RotatableStreamDestination.php:37]: Multiple inheritance is not possible, and you want to be 5.3 compatible, so I can not use Traits. So this is little bit redundant code - for any new implementation of destination, same 'preRotator' / 'postRotator' logic would have to be implemented. For now, we are fine... we will see in the future how to handle this.... I was thinking about Gaufrete as Destination driver implementation as well...

File: Destination/RotatableStreamDestination.php [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+
+    /**
+     * Remove backup from backup location.
+     *
+     * @param Backup $backup Backup to remove.
+     * @param LoggerInterface $logger A logger.
+     */
+    private function remove(Backup $backup, LoggerInterface $logger)
```
Comment (TheCelavi) [inline Destination/RotatableStreamDestination.php]: Remove is now private, ok.

File: Destination/StreamDestination.php [hunk 11]
```diff
@@ -23,6 +32,53 @@ public function push($filename, LoggerInterface $logger)
     {
         $logger->info(sprintf('Copying %s to %s', $filename, $this->directory));
 
-        copy($filename, sprintf('%s/%s', $this->directory, basename($filename)));
+        copy($filename, ($destination = sprintf('%s/%s', $this->directory, basename($filename))));
+        touch($destination, filemtime($filename));
```
Comment (TheCelavi) [inline Destination/StreamDestination.php:35]: This 'touch' is very important!!! Rotation can not work properly without it!

File: Rotator/MaxStorageSizeRotator.php [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+     * A constructor.
+     *
+     * @param integer|string $maxSize Maximum storage size.
+     */
+    public function __construct($maxSize)
+    {
+        if (is_numeric($maxSize)) {
+            $this->maxSize = $maxSize;
+        } else {
+            if (class_exists('\\ByteUnits\\System')) {
```
Comment (TheCelavi) [inline Rotator/MaxStorageSizeRotator.php:25]: Ok, there is optional usage of \ByteUnits, but we will fallback to our own implementation, because common use case: 100GB, 10GB, 500MB -> it is really easy to implement.

File: Rotator/MinCountMaxStorageSizeRotator.php [hunk 13]
```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Zenstruck\BackupBundle\Rotator;
+
+use Zenstruck\BackupBundle\Destination\Backup;
+use Zenstruck\BackupBundle\Utils\Filesize;
+
+final class MinCountMaxStorageSizeRotator
```
Comment (TheCelavi) [inline Rotator/MinCountMaxStorageSizeRotator.php:8]: New implementation of rotator -> max size but keep minimum number of backups regardless of size.

File: Rotator/RotatorCollection.php [hunk 14]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace Zenstruck\BackupBundle\Rotator;
+
+use ArrayIterator;
+use Countable;
+use IteratorAggregate;
+
+final class RotatorCollection implements Countable, IteratorAggregate, Rotator
```
Comment (TheCelavi) [inline Rotator/RotatorCollection.php]: Ok, this is rotator collection, so pre/post rotations can be decided with multiple rotators.

File: Tests/Destination/RotatableStreamDestinationTest.php [hunk 15]
```diff
@@ -0,0 +1,87 @@
+<?php
```
Comment (TheCelavi) [inline Tests/Destination/RotatableStreamDestinationTest.php:1]: Ok, I have wrote some tests as well...

File: Tests/Fixtures/NullDestination.php [hunk 16]
```diff
@@ -17,4 +18,20 @@ public function push($filename, LoggerInterface $logger)
     {
         // noop
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getIterator()
```
Comment (TheCelavi) [inline Tests/Fixtures/NullDestination.php:25]: In order to pass test, these methods had to be added to NullDestination.

File: Tests/Fixtures/rotation/1.txt [hunk 17]
```diff
@@ -0,0 +1 @@
+foo
```
Comment (TheCelavi) [inline Tests/Fixtures/rotation/1.txt:1]: These files, 1.txt to 5.txt are needed for testing rotations.

File: Utils/Filesize.php [hunk 18]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+namespace Zenstruck\BackupBundle\Utils;
+
+abstract class Filesize
+{
+    private final function __construct() {}
+
+    public static function getBytes($size)
```
Comment (TheCelavi) [inline Utils/Filesize.php:9]: This is small implementation for common pretty size format, there is really no need for dependency....

File: phpunit.xml.dist [hunk 19]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 
-<phpunit bootstrap="./vendor/autoload.php" colors="true">
+<phpunit colors="true">
```
Comment (TheCelavi) [inline phpunit.xml.dist]: I have removed "bootstrap" declaration from dist file. 

I use IntelliJ for testing, and it is easy to setup all custom paths - to PhpUnit, to Composer, to phpunit.xml...  So this declaration is not required at all.

Moreover -> I would like to restructure bundle:

-> root
   -> src (bundle code)
   -> tests (tests)
   -> other files, composer.json, readme.md and so on...

And I would like to add 'autoload-dev' declaration in composer.json, see per example: https://github.com/RunOpenCode/assets-injection/blob/master/composer.json

Comment (kbond) [inline phpunit.xml.dist]: The bootstrap is required for travis. Moreover, we want people to be able to run the tests by just running phpunit from the root - it should be IDE agnostic.

For the bundle structure, I agree, but let's do that in another PR after this is merged.

Comment (TheCelavi) [inline phpunit.xml.dist]: Understood, will revert that ASAP.

File: Destination/Destination.php (hunk 1 above)
Comment (kbond) [inline Destination/Destination.php]: I would like the destination interface to remain as is. There is no reason for `S3CmdDestinaion` to implement `Countable`, `IteratorAggregate`. If you want `RotatingStreamDestination` to implement them, that's fine.

Comment (TheCelavi) [inline Destination/Destination.php]: How to implement GUI over this if there is no way to iterate over the Destination? How to get list of current backups?

Comment (kbond) [inline Destination/Destination.php]: Ok, I see what you mean.

If you are building a GUI, do you not want the functionality to remove backups as well? We should think about what we want the GUI to do. That functionality might be out the scope of the `Destination`. I intended the destination to only push a file.

For instance, the S3CmdDestination can really never do more than push - it can't list or remove (well i suppose it could but it would be exceedingly difficult).

I think we need a separate system for "administrating" the backups. Maybe even a separate bundle.

File: Destination/RotatableStreamDestination.php (hunk 9 above)
Comment (kbond) [inline Destination/RotatableStreamDestination.php:37]: Maybe what we need is another interface which extends `Destination` that adds methods like list/remove and it could be Countable/IteratorAggregate. Something like:

``` php
interface DestinationExtra extends Destination, \Countable, \IteratorAggregate
{
    /**
     * @return Backup[]
     */
    public function list();

    public function remove(Backup $backup);
}
```

Can't think of a good name for the class, `DestinationExtra` is not great.

File: Destination/Destination.php (hunk 1 above)
Comment (TheCelavi) [inline Destination/Destination.php]: It is not only about GUI -> if destination can not be iteratable/readable, than it is a black hole that swallows backups only -> we can not list files, see what is in it, what is the amount of occupied space… etc…. 

In regards to S3Cmc -> according to the documentation, it is possible to navigate trough buckets, to list files, get file infos, etc…, see: http://s3tools.org/usage http://s3tools.org/usage

It would be quite unusual that it is impossible to get details about the storage…

GUI should be separate bundle, I do not mind that -> but I really believe that Destination should be iteratable and should allow to the developer to list trough all current stored backups. 

What is the issue having Destination implementing possibility to iterate trough current backups - what are the risks, side effects, dangers?

File: Destination/RotatableStreamDestination.php (hunk 9 above)
Comment (kbond) [inline Destination/RotatableStreamDestination.php:37]: I also like the idea of Gaufrette/Flysystem Destinations.

Comment (TheCelavi) [inline Destination/RotatableStreamDestination.php:37]: I have to disagree on that, I do not see the reason why there should be a separate interface.

Think of Destination as a bag of stuffs. You put stuffs in there. Is that end of the story?
- Would you like to peak into bag into one point of time?
- Would you like to remove old, rotten stuff from bag?
- Would you like to know how many thing you have in the bag?
- Would you like to take something from bag and give that to someone?

What is a bag if only thing that you can do with it is to put stuff in it?

Ok, if user should not be allowed to remove backup - that is fine, no "remove" method, but Iteration is quite natural thing to have.

I have stated a lot of arguments that supports the necessity of iterating trough backups in one destination. I would like to hear counter-arguments and then I can understand your position.

File: Destination/Destination.php (hunk 1 above)
Comment (kbond) [inline Destination/Destination.php]: Yes, it is possible to do that with S3Cmd but we would have to parse the output of the command which would get ugly.

For my use case, I want it to be a black hole. I have policies on S3 that manage the backups for me (moving old ones to Glacier and deleting the really old ones). For me, the backup is fire and forget.

There is no risk/danger but for some destinations (like S3Cmd) it doesn't make sense.

Comment (kbond) [inline Destination/Destination.php]: Ok, I looked at the output of `s3cmd ls <bucket>/prefix/` and it isn't as bad as I thought. Should be able to parse it easy enough with a bit of regex.

So, let's go with having the Destination Countable/IteratorAggregate.

I still would like `remove`/`list` methods.

Comment (kbond) [inline Destination/Destination.php]: err, actually, just a `remove` method

Comment (TheCelavi) [inline Destination/Destination.php]: Ok, lets have those.

'remove()' is possible, list is not, it is reserved, it has to be either 'getAll()' or 'getBackups()'.

Do you agree to have both -> Iterator/Countable + getBackups() => it is practically the same.

Comment (TheCelavi) [inline Destination/Destination.php]: Ok, only "remove(Backup $backup)" it is.

File: Destination/StreamDestination.php [hunk 20]
```diff
@@ -23,6 +32,53 @@ public function push($filename, LoggerInterface $logger)
...
+            $this->lazyLoadBackups();
+        }
+
+        return count($this->backups);
+    }
+
+    /**
+     * Loads list of backups in lazy manner.
+     */
+    protected function lazyLoadBackups()
```
Comment (kbond) [inline Destination/StreamDestination.php]: Instead of having `is_array` checks all over the place, just have:

``` php
protected function getBackups()
{
    if (null !== $this->backups) {
        return $this->backups;
    }

    // ... contents of lazyloadbackups
}
```

That way you can just call `$this->getBackups()` any time you need them.

Comment (TheCelavi) [inline Destination/StreamDestination.php]: Agree.

File: Destination/StreamDestination.php [hunk 21]
```diff
@@ -9,8 +10,16 @@
  */
 class StreamDestination implements Destination
 {
+    /**
+     * @var string Path to directory.
+     */
     private $directory;
 
+    /**
+     * @var array<Backup> List of backups.
```
Comment (kbond) [inline Destination/StreamDestination.php:18]: Can you change to `Backup[]` instead? Admittedly, I like your way better but the rest of the bundle uses `Class[]`.

File: Destination/StreamDestination.php [hunk 22]
```diff
@@ -2,6 +2,7 @@
 
 namespace Zenstruck\BackupBundle\Destination;
 
+use ArrayIterator;
```
Comment (kbond) [inline Destination/StreamDestination.php]: For global namespaces, can you just inline them instead of adding the use.

File: Destination/StreamDestination.php (hunk 21 above)
Comment (TheCelavi) [inline Destination/StreamDestination.php:18]: Of course.

File: Destination/StreamDestination.php (hunk 22 above)
Comment (TheCelavi) [inline Destination/StreamDestination.php]: Ok.

File: Destination/Destination.php (hunk 1 above)
Comment (kbond) [inline Destination/Destination.php]: What about a `getBackup($key)` method. `getIterator` should return an array keyed by the Backup key.

I'm thinking in a UI you would have a list of backups and a "delete" button. You would need an ID.

Since the key is probably a path maybe you can md5 it?

Comment (TheCelavi) [inline Destination/Destination.php]: Agree.

File: Destination/Backup.php [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    public function getFilename()
+    {
+        return $this->filename;
+    }
+
+    /**
+     * @return int
+     */
+    public function getSize()
```
Comment (kbond) [inline Destination/Backup.php:62]: Can we have a `getHumanReadable()` method?

Also, I think we should use https://github.com/gabrielelana/byte-units library for converting.

Comment (TheCelavi) [inline Destination/Backup.php:62]: I don’t think that that is good idea, on View layer (wether is that Twig template, or maybe console application) formatting of file size should be done by some sort of template function (Twig filter).

Class Backup belongs to Model layer, and it should not be in charge for data representation, it will violate Single responsibility princip.

Comment (kbond) [inline Destination/Backup.php:62]: Ok, fair point.

I still think we should use the 3rd party library for parsing the $size from config rather than a Util class.

Comment (TheCelavi) [inline Destination/Backup.php:62]: Me to, but people tends to avoid dependencies, so therefore - both are used - it will try library that you have provided, and if it can not be found, it will fall back to Util class.

In composer.json we will put library that you recommended under suggest section, and state that in documentation.

Util class will cover use cases like, 100MB, 100GB, 1TB when configuring rotators -> which is 99%. of use cases. 

You can not make someone to pull whole library for only one functionality which can fit into one method. :)

**Comment by TheCelavi** [comment]
Ok, all done. 

If everything is fine - I have to leave you with implementation of S3CmdDestination - I do not have account to test that... sorry :( 

Now, do review, and notify me when I can start to mess around with factories and integration of rotators into service container.

Thanks for your time.

File: Destination/Backup.php [hunk 24]
```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace Zenstruck\BackupBundle\Destination;
+use DateTime;
```
Comment (kbond) [inline Destination/Backup.php:4]: Can you inline this?

File: Destination/S3CmdDestination.php [hunk 25]
```diff
@@ -54,4 +54,45 @@ function ($type, $buffer) use ($logger) {
...
+    {
+        throw new \RuntimeException('Not implemented yet');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function remove(Backup $backup, LoggerInterface $logger)
+    {
+        // TODO: Implement remove() method.
```
Comment (kbond) [inline Destination/S3CmdDestination.php:80]: Can you add a \RuntimeException here?

File: Destination/StreamDestination.php [hunk 26]
```diff
@@ -23,6 +31,91 @@ public function push($filename, LoggerInterface $logger)
...
+    /**
+     * {@inheritdoc}
+     */
+    public function has($key)
+    {
+        return array_key_exists($key, $this->getBackups());
+    }
+
+    /**
+     * Get backups.
```
Comment (kbond) [inline Destination/StreamDestination.php:96]: remove this.

File: Rotator/MinCountMaxStorageSizeRotator.php (hunk 13 above)
Comment (kbond) [inline Rotator/MinCountMaxStorageSizeRotator.php:8]: Should this not implement `Rotator`?

File: Tests/Rotator/MinCountMaxStorageSizeRotatorTest.php [hunk 27]
```diff
@@ -0,0 +1,54 @@
+<?php
+/**
```
Comment (kbond) [inline Tests/Rotator/MinCountMaxStorageSizeRotatorTest.php:2]: can you remove these comments?

File: Destination/StreamDestination.php (hunk 21 above)
Comment (kbond) [inline Destination/StreamDestination.php:18]: ^^

File: Rotator/MaxCountRotator.php [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+
+final class MaxCountRotator implements Rotator
+{
+    /**
+     * @var integer
+     */
+    private $count;
+
+    /**
+     * A constructor.
```
Comment (kbond) [inline Rotator/MaxCountRotator.php:15]: This comment isn't needed.

File: Rotator/MaxStorageSizeRotator.php [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+
+final class MaxStorageSizeRotator implements Rotator
+{
+    /**
+     * @var integer
+     */
+    private $maxSize;
+
+    /**
+     * A constructor.
```
Comment (kbond) [inline Rotator/MaxStorageSizeRotator.php:16]: Again, not needed.

File: Rotator/MinCountMaxStorageSizeRotator.php [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    public function __construct($minCount, $maxSize)
+    {
+        if ($minCount < 1) {
+            throw new \InvalidArgumentException('At least one file should be set as minimum backup count.');
+        }
+
+        $this->minCount = $minCount;
+
+        if (is_integer($maxSize)) {
```
Comment (kbond) [inline Rotator/MinCountMaxStorageSizeRotator.php:34]: This is kind of ugly, can you add a private method?

``` php
$this->maxSize = $this->calculateMaxsize($maxSize);
```

File: Rotator/MaxStorageSizeRotator.php [hunk 31]
```diff
@@ -0,0 +1,74 @@
...
+     *
+     * @param integer|string $maxSize Maximum storage size.
```
Comment (kbond) [inline Rotator/MaxStorageSizeRotator.php:18]: Can you typehint `integer` with just `int`

File: Rotator/MinCountMaxStorageSizeRotator.php [hunk 32]
```diff
@@ -0,0 +1,90 @@
...
+
+        /**
+         * @var Backup $backup
+         */
+        foreach ($backups as $backup) {
+            $list[$backup->getCreatedAt()->getTimestamp()] = $backup;
+            $currentSize += $backup->getSize();
+        }
+
+        if ($currentSize > $this->maxSize) {
```
Comment (kbond) [inline Rotator/MinCountMaxStorageSizeRotator.php:61]: can you change to remove the `else`:

``` php
if ($currentSize <= $this->maxSize) {
    return array();
}

// ...
```

File: Rotator/MinCountMaxStorageSizeRotator.php (hunk 30 above)
Comment (kbond) [inline Rotator/MinCountMaxStorageSizeRotator.php:34]: Also, this is duplicated code - perhaps move to `Util\Filesize`?

File: Destination/Destination.php [hunk 33]
```diff
@@ -7,11 +7,38 @@
...
+     *
+     * @param Backup $backup
+     * @param LoggerInterface $logger A logger.
+     */
+    public function remove(Backup $backup, LoggerInterface $logger);
+
+    /**
+     * Get backup with given key.
+     *
+     * @param string $key Backup key.
```
Comment (kbond) [inline Destination/Destination.php:31]: Can you add a newline between `@param(s)`, `@return` and `@throws`

File: Destination/RotatableStreamDestination.php [hunk 34]
```diff
@@ -0,0 +1,75 @@
...
+            foreach ($nominations as $nomination) {
+                $this->doRotate($nomination, $logger);
+            }
+        }
+    }
+
+    /**
+     * Remove backup from backup location due to rotation.
+     *
+     * @param Backup $backup Backup to remove.
```
Comment (kbond) [inline Destination/RotatableStreamDestination.php:67]: can you line the variables up?

``` php
/**
 * Remove backup from backup location due to rotation.
 *
 * @param Backup          $backup Backup to remove.
 * @param LoggerInterface $logger A logger.
 */
```

File: Destination/StreamDestination.php [hunk 35]
```diff
@@ -23,6 +31,91 @@ public function push($filename, LoggerInterface $logger)
...
+    {
+        return count($this->getBackups());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function remove(Backup $backup, LoggerInterface $logger)
+    {
+        @unlink($backup->getFilename());
```
Comment (kbond) [inline Destination/StreamDestination.php:66]: You should check if this failed and throw an exception.

**Comment by kbond** [comment]
Looks good except for a few CS issues, duplicated code and cleanup. I try not using the `else` statement and nested `for` loops if I can avoid it.

**Comment by kachkaev** [comment]
Hi guys,

Am I right that this PR will eventually enable rotations of backups? My ultimate goal is to have hourly backups over the last day, daily backups over the last week and weekly backups for anything that's older. This means that the backup command will have to clean-up some archives that exist at the destination (or each destination, if there are more than one).

Is this what's going to be possible? How soon are you planning to roll out the change?

I'm currently choosing between kbond/ZenstruckBackupBundle and  dizda/CloudBackupBundle, your answer will help a lot with the decision.

**Comment by kbond** [comment]
@kachkaev I believe this PR will enable what you desire.  @TheCelavi would be able to better answer though.

**Comment by kbond** [comment]
@TheCelavi I am in the process of moving the backup specific stuff to it's own repo (https://github.com/kbond/php-backup). When I have some time, I can try my hand at migrating what you have done here over.

**Comment by TheCelavi** [comment]
Hi, I am sorry, I have some much stuff on my plate right now so I am not in the position to dedicate the time to this topic right now, sorry. :(

**Comment by TheCelavi** [comment]
You have the code, so if you want you can adjust it to your needs, I am going to close the PR since I do not have time right now to handle it.
